# Notes for DE analsysis of the different comparisons

## Outline
1. Identify General and Time-dependent DE genes using DESeq2
2. Identify General and Time-dependent DE genes using maSigPro
3. Generate contrast-specific lists of DE genes that combine results from 
the two methods
  * General and Time-dependent
4. Generate modules of gene co-expression using multiple contrasts

## DESeq2
### Overview
#### My Notes
Uses negative binomial generalized linear models and estimated dispersion and
log-fold-changes for different parameters in the model
#### Resources
##### Bioconductor Outline
I followed the outline provided by Bioconductor:
http://master.bioconductor.org/packages/release/workflows/vignettes/rnaseqGene/inst/doc/rnaseqGene.html
##### Package Website
https://bioconductor.org/packages/release/bioc/html/DESeq2.html
##### Manual
https://bioconductor.org/packages/release/bioc/manuals/DESeq2/man/DESeq2.pdf
#### Explanation of Files Generated by Scripts
Returns 2 groups of 3 files:
##### Groups
* `<general>`: genes that are different between the Control and Treatment after
accounting for the other parameters, specifically, time.
These genes "overall" differences and not necessarily different
responses/reactions across time.
Usually, this will be a large list.
* `<TC>`: genes that show differences at different time points that are
different from the "global" differences between Control and Treatment.
For instance: Expression in Treatment line increases from T0 to T1 but is
unchanged in Control line.
##### Files Within Groups
* `<PREFIX_DESeq2_GROUP_genes.txt>`: contains all genes, overall p-value,
and multiple testing-corrected p-value
* `<PREFIX_DESeq2_TC_var_lfc.txt>`: (only for TC group) contains the estimated
log-fold-change associated with each variable in the model (not including
TimeVsTime variables). This can be used for trying to cluster genes based on
their modeled response across conditions
* `<PREFIX_DESeq2_GROUP_full_mat.txt>`: contains the gene name, average
expression parameter values, average expression, F value, overall p-value, and
multiple testing-adjusted p-value for the significant genes
### Identify Differentially Expressed (DE) genes
* Original R script uses two approaches in DESeq2 (general and TC) to identify
differentially expressed genes
* Adapted R script only lookds for general DE genes if there is only one time 
point in the comparison
#### R Scripts
* General and TC 
  * `~/CamSat_transcript_comps/r_scripts/DESeq2_script.r`
* Just General (no TC)
  * `~/CamSat_transcript_comps/r_scripts/DESeq2_script_1timepoint.r`
#### Submit job for TC comparisons on Cori
* Included all the comparisons in a single job
```
cd /global/cscratch1/sd/grabowsp/CamSat_transcript/Cs_transc_DESeq2
sbatch run_DESeq2_01.sh
```
#### Run MT5 vs HMT102 interactively on Cori
* Only 1 time point, so don't need the TC component
```
cd /global/cscratch1/sd/grabowsp/CamSat_transcript/Cs_transc_DESeq2

module load python/3.7-anaconda-2019.07
source activate R_RNAseq

Rscript \
/global/homes/g/grabowsp/tools/CamSat_transcript_comps/r_scripts/\
DESeq2_script_1timepoint.r \
/global/cscratch1/sd/grabowsp/CamSat_transcript/Cs_transc_counts/\
Camelina_transc_tot_counts_noBadLibs.txt \
/global/cscratch1/sd/grabowsp/CamSat_transcript/Cs_rnaseq_meta_for_DESeq2.txt \
MT5_vs_HMT102 MT5 HMT102 15DAF

```

## maSigPro
### Overview
#### My Notes
Fits regression including different parameters for relationship with time
(ex: linear, quadratic), then tests how different the fit is when parameters
are modeled separately for the treatment vs modeled for treatment and control
together.
Then goes back and removed parameters to test which parameters are important
for the difference.
#### Resources
##### maSigPro Website
https://www.bioconductor.org/packages/release/bioc/html/maSigPro.html
##### maSigPro Bioconductor Users Guide
https://www.bioconductor.org/packages/release/bioc/vignettes/maSigPro/inst/doc/maSigProUsersGuide.pdf
##### maSigPro Reference Manual
https://www.bioconductor.org/packages/release/bioc/manuals/maSigPro/man/maSigPro.pdf
#### Explanation of Files Generated by Scripts
Returns 3 files:
* `<PREFIX_DE_genes.txt>`: contains all genes and overall p-value
(corresponding to the difference across all parameters)
* `<PREFIX_DE_betas.txt>`: contains the gene names and the betas for all the
parameters, for the significant genes.
This can be used for trying to cluster genes based on their modeled response
across conditions
* `<PREFIX_DE_full_mat.txt>`: contains the gene name, overall p-value, R^2,
and beta and p-value for each parameter for the significant genes
### Identify DE Genes
#### R Scripts
* `~/CamSat_transcript_comps/r_scripts/maSigPro_TC_script.r`
#### Run R script 
```
cd /global/cscratch1/sd/grabowsp/CamSat_transcript/Cs_transc_maSigPro

sbatch run_maSigPro_01.sh
# because of Cori queue waiting, I had to run the command during interactive
#  session

## initially ran this below, but had to re-run everything...
# re-run some of the comparisons using an updated r-script that corrects
#   an error when using '-' in labels of treatments
#sbatch run_maSigPro_02.sh
```

## Make combined lists
### Overview
* Make lists of the genes that overlap for either "General" or "TC" in \
DESeq2 and maSigPro results
### Make General Combined Lists
```
module load python/3.7-anaconda-2019.07
source activate R_GO_analysis

Rscript /global/homes/g/grabowsp/tools/CamSat_transcript_comps/r_scripts/make_General_DE_List1.r
```
